[WIP] RD Training (#57521)
odoo/odoo: doc/howtos/rdtraining/15_qwebintro.rst

.. _howto/rdtraining/15_qwebintro:

================================
Part 15: A Brief History Of QWeb
================================

> Up to now, the design of the interface of our real estate module was rather limited. Building

Reviewer: ```suggestion
Up to now, the design of the interface of our real estate module was rather rigid. Building
```
or "simple".

a list view is straightforward since only the list of fields is necessary. The same holds true
for the form view: despite the use of a few tags such as ``<group>`` or ``<page>``, there
is very little to do in terms of design.

> However, if we want to give a unique look to our application, it is necessary to go a step

Reviewer: And who wouldn't?
![382f4a83761e875823f30adf77308cad](https://user-images.githubusercontent.com/7571158/96463588-2775fb00-1227-11eb-84fc-ed30d1d58a73.jpg)

further and be able to design new views. Moreover, other features such as PDF reports or
website pages need another tool to be created with more flexibility: a templating_ engine.

You might already be familiar with existing engines such as Jinja (Python), ERB (Ruby) or
Twig (PHP). Odoo comes with its own built-in engine: :ref:`reference/qweb`.
QWeb is the primary templating engine used by Odoo. It is an XML templating engine and used
mostly to generate HTML fragments and pages.

You probably already came across the `kanban board`_ in Odoo: the records are displayed in a
card-like structure. We will build such a view for our real estate module.

Concrete Example: A Kanban View
===============================

**Reference**: the documentation related to this topic can be found in
:ref:`reference/views/kanban`.

.. note::

    **Goal**: at the end of this section a Kanban view of the properties is created:

    .. image:: 15_qwebintro/media/kanban.png
        :align: center
        :alt: Kanban view

In our estate application, we would like to add a Kanban view to display our properties. Kanban

> views are standard Odoo views (such as the form or list view), but their structure is much more

Reviewer: ```suggestion
views are standard Odoo views (similar to the form or list views), but their structure is more
```
Form views can include pretty arbitrary HTML so they're quite flexible.

flexible. In fact, the structure of each card is a mix of form elements (including basic HTML)
and QWeb. The definition of a Kanban view is similar to the definition of the list and form
views, except that their root element is ``<kanban>``. In its simplest form, a Kanban view
looks like:

.. code-block:: xml

    <kanban>
        <templates>
            <t t-name="kanban-box">
                <div class="oe_kanban_global_click">
                    <field name="name"/>
                </div>
            </t>
        </templates>
    </kanban>

Let's break down this example:

- ``<templates>``: defines a list of :ref:`reference/qweb` templates. Kanban views *must* define at

>   least one root template ``kanban-box``, which will be rendered once for each record.

Reviewer: Explain what other templates we'd define? Maybe show / link to an ornate kanban which uses multiple templates?

    Tip: you must add ``kanban`` in the ``view_mode`` of the corresponding
    ``ir.actions.act_window``.

Once the Kanban view is working, we can start improving it. If we want to display an element
conditionally, we can use the ``t-if`` directive (see :ref:`reference/qweb/conditionals`).

.. code-block:: xml

    <kanban>
        <field name="state"/>
        <templates>
            <t t-name="kanban-box">
                <div class="oe_kanban_global_click">
                    <field name="name"/>
                </div>
                <div t-if="record.state.raw_value == 'new'">
                    This is new!
                </div>
            </t>
        </templates>
    </kanban>

We added a few things:

- ``t-if``: the ``<div>`` element is rendered if the condition is true.
- ``record``: an object with all the requested fields as its attributes. Each field has
  two attributes ``value`` and ``raw_value``, the former is formatted according to current
  user parameters, the latter is the direct value from a :meth:`~odoo.models.Model.read`.

In the above example, the field ``name`` was added in the ``<templates>`` element, but ``state``

> is outside. When we need the value of a field but not display it in the view, it is possible to

Reviewer: Maybe clarify / expand on this e.g. "the kanban view fetches as little as it can so the value isn't even on the pseudo-record unless we request it"?
